ion/unmarshal.go

```diff
@@ -0,0 +1,673 @@
+package ion
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+var (
+	// ErrNoInput is returned when there is no input to decode
+	ErrNoInput = errors.New("ion: no input to decode")
+)
+
+// Unmarshal unmarshals Ion data to the given object.
+func Unmarshal(data []byte, v interface{}) error {
+	return NewDecoder(NewReader(bytes.NewReader(data))).DecodeTo(v)
+}
+
+// UnmarshalStr unmarshals Ion data from a string to the given object.
+func UnmarshalStr(data string, v interface{}) error {
+	return Unmarshal([]byte(data), v)
+}
+
+// UnmarshalFrom unmarshal Ion data from a reader to the given object.
+func UnmarshalFrom(r Reader, v interface{}) error {
+	d := Decoder{
+		r: r,
+	}
+	return d.DecodeTo(v)
+}
+
+// A Decoder decodes go values from an Ion reader.
+type Decoder struct {
+	r Reader
+}
+
+// NewDecoder creates a new decoder.
+func NewDecoder(r Reader) *Decoder {
+	return &Decoder{
+		r: r,
+	}
+}
+
+// NewTextDecoder creates a new text decoder. Well, a decoder that uses a reader with
+// no shared symbol tables, it'll work to read binary too if the binary doesn't reference
+// any shared symbol tables.
+func NewTextDecoder(in io.Reader) *Decoder {
```

Comment:
I don't think that the decoder story is clear.  What if I have binary data that is composed of multiple Ion datagrams each one should decode into an instance of a struct?  What is the binary story?  If the reader is supposed to be able to determine text or binary, then why are there multiple ways to create a Decoder?

Reply:
Decoders read one value at a time, without respect to datagrams (the symbol table gets reset under the covers when we hit a new BVM, but that isn't exposed up through the Reader). I didn't have a use case for decoding an entire datagram at a time, if you do please feel free to refactor to support that.

`NewTextDecoder` is a  convenience since I was mostly decoding text Ion and got sick of typing `NewDecoder(NewReader(in))`. :) It's not well-named, feel free to change or delete it.

Reply:
The Encode/Decode as well as the Marshall/Unmarshall are a good seed and is already on our todo list.